File: src/typescript/emits-events.ts
import { StructuredEvent, StreamConfig } from "./event-stream-parser.js";
import { MessageBus } from "./message-bus.js";

export { StructuredEvent, StreamConfig };

type Constructor<T = {}> = new (...args: any[]) => T;

export function EmitsEvents<TBase extends Constructor>(Base: TBase) {
  return class extends Base {
    eventRegistry: Map<string, StructuredEvent> = new Map();
    defaultEvents: string[] = [];
    private _bus: MessageBus = new MessageBus();

    registerEvent(eventType: StructuredEvent): void {
      this.eventRegistry.set(eventType.name, eventType);
    }

    resolveActiveEvents(events?: (string | StructuredEvent)[]): StructuredEvent[] {
      if (events === undefined) {
        return this.defaultEvents
          .filter((name) => this.eventRegistry.has(name))
          .map((name) => this.eventRegistry.get(name)!);
      }
      const result: StructuredEvent[] = [];
      for (const item of events) {
        if (typeof item === "string") {
          const et = this.eventRegistry.get(item);
          if (et) result.push(et);
        } else {
          result.push(item);
        }
      }
      return result;
    }

    buildEventPrompt(eventTypes: StructuredEvent[]): string {
      if (eventTypes.length === 0) return "";
      const sections: string[] = [];
      sections.push("# Event Emission");
      sections.push("");
      sections.push(
        "You can emit structured events inline in your response using the following format:"
      );
      sections.push("");

      for (const et of eventTypes) {
        sections.push(`## Event: ${et.name}`);
        sections.push(`Description: ${et.description}`);
        sections.push("Format:");
        sections.push("```");
        sections.push("---event");
        sections.push(`type: ${et.name}`);
        if (et.schema) {
          for (const [key, val] of Object.entries(et.schema)) {
            if (typeof val === "object" && val !== null && !Array.isArray(val)) {
              sections.push(`${key}:`);
              for (const [k, v] of Object.entries(val)) {
                sections.push(`  ${k}: <${v}>`);
              }
            } else {
              sections.push(`${key}: <${val}>`);
            }
          }
        }
        sections.push("---");
        sections.push("```");
        if (et.instructions) {
          sections.push(et.instructions);
        }
        sections.push("");
      }

      return sections.join("\n");
    }

    get bus(): MessageBus {
      return this._bus;
    }

    set bus(value: MessageBus) {
      this._bus = value;
    }
  };
}

File: src/python/emits_events.py
from __future__ import annotations

from typing import Any, Self

from src.python.event_stream_parser import EventType, StructuredEvent, StreamConfig
from src.python.message_bus import MessageBus


class EmitsEvents:
    def __init_emits_events__(self) -> None:
        self._event_registry: dict[str, EventType] = {}
        if not hasattr(self, "default_events"):
            self.default_events: list[str] = []
        self._bus: MessageBus = MessageBus()

    def _ensure_events_init(self) -> None:
        if not hasattr(self, "_event_registry"):
            self.__init_emits_events__()

    def register_event(self, event_type: EventType) -> Self:
        self._ensure_events_init()
        self._event_registry[event_type.name] = event_type
        return self

    def unregister_event(self, name: str) -> Self:
        self._ensure_events_init()
        self._event_registry.pop(name, None)
        return self

    @property
    def events(self) -> dict[str, EventType]:
        self._ensure_events_init()
        return dict(self._event_registry)

    def _resolve_active_events(
        self, events: list[str | EventType] | None = None
    ) -> list[EventType]:
        self._ensure_events_init()
        if events is None:
            return [
                self._event_registry[name]
                for name in self.default_events
                if name in self._event_registry
            ]
        result: list[EventType] = []
        for item in events:
            if isinstance(item, str):
                if item in self._event_registry:
                    result.append(self._event_registry[item])
            elif isinstance(item, EventType):
                result.append(item)
        return result

    def _build_event_prompt(self, event_types: list[EventType]) -> str:
        if not event_types:
            return ""
        sections: list[str] = []
        sections.append("# Event Emission")
        sections.append("")
        sections.append("You can emit structured events inline in your response using the following format:")
        sections.append("")

        for et in event_types:
            sections.append(f"## Event: {et.name}")
            sections.append(f"Description: {et.description}")
            sections.append("Format:")
            sections.append("```")
            sections.append("---event")
            sections.append(f"type: {et.name}")
            if et.schema:
                for key, val in et.schema.items():
                    if isinstance(val, dict):
                        sections.append(f"{key}:")
                        for k, v in val.items():
                            sections.append(f"  {k}: <{v}>")
                    else:
                        sections.append(f"{key}: <{val}>")
            sections.append("---")
            sections.append("```")
            if et.instructions:
                sections.append(et.instructions)
            sections.append("")

        return "\n".join(sections)

    @property
    def bus(self) -> MessageBus:
        self._ensure_events_init()
        return self._bus

    @bus.setter
    def bus(self, value: MessageBus) -> None:
        self._ensure_events_init()
        self._bus = value
